src/fp/list.nim

```diff
@@ -114,6 +114,24 @@ proc foldRightF*[T, U](xs: List[T], z: () -> U, f: (T, () -> U) -> U): U =
   if xs.isEmpty: z()
   else: f(xs.head, () => xs.tail.foldRightF(z, f))
 
+proc unfoldLeft*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the left from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = b ^^ result
+
+proc unfoldRight*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the right from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = result.append(b.asList)
```

Comment:
Indeed, after the PR, I was worried about the performance of appending to a cons-List.

Besides reversing as suggested which traverses the list twice, the only solution I found so far was in Lisp on this StackOverflow [post](https://stackoverflow.com/questions/6439972/what-is-the-cons-to-add-an-item-to-the-end-of-the-list), which is too cryptic for me (for now)

```Lisp
(defun track-tail (count)
  (let* ((list (cons 0 nil))
     (tail list))
    (loop for n from 1 below count
       do (progn
        (setf (cdr tail) (cons n nil))
        (setf tail (cdr tail))
        (format t "With n == ~d, the list is ~a~%" n list)))
    list))
```
Output
```
CL-USER> (track-tail 5)
With n == 1, the list is (0 1)
With n == 2, the list is (0 1 2)
With n == 3, the list is (0 1 2 3)
With n == 4, the list is (0 1 2 3 4)
(0 1 2 3 4)
```

For now I will use the reversing solution